<launch>

	<!-- Arguments -->
	<arg name="bagfile" default="/home/echo/Desktop/ugv_slam_data/kitti_data_odometry_gray_sequence_00.bag"/>
  <arg name="camera" default="/kitti"/>
  <param name="/use_sim_time" value="true"/>

  <arg name="approximate_sync"      default="true" />
  <arg name="world_frame_id"        default="world" />
  <arg name="base_link_frame_id"    default="base_link" />
  <arg name="gps_frame_id"          default="gps" />
  <arg name="sensor_frame_id"       default="camera_gray_left" />
  <arg name="is_kitti"              default="true" />
  <arg name="add_gps_noise"         default="true" />

  <!-- Play the bagfile, note the rename of topics-->
  <node pkg="rosbag" type="play" name="rosbag" 
        args="--pause --clock $(arg bagfile)
              /kitti/camera_gray_left/camera_info:=/kitti/left/camera_info
              /kitti/camera_gray_right/camera_info:=/kitti/right/camera_info
              /kitti/camera_gray_left/image_rect:=/kitti/left/image_rect
              /kitti/camera_gray_right/image_rect:=/kitti/right/image_rect" />

  <node pkg="tf" type="static_transform_publisher" name="tf_base2cam" args="0 0 0 0 0 0 1 $(arg base_link_frame_id) $(arg sensor_frame_id) 100" />
  <node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc" ns="$(arg camera)" />

  <!-- Viso2 -->
  <node pkg="viso2_slam" type="stereo_slam" name="stereo_slam" output="screen">
    <remap from="stereo" to="$(arg camera)"/>
    <remap from="image" to="image_rect"/>
    <!-- Common params -->
    <param name="odom_frame_id"         value="$(arg world_frame_id)"/>
    <param name="base_link_frame_id"    value="$(arg base_link_frame_id)"/>
    <param name="publish_tf"            value="true"/>
    <!-- Stereo params -->
    <param name="approximate_sync"      value="$(arg approximate_sync)"/>
    <param name="sensor_frame_id"       value="$(arg sensor_frame_id)"/>
  </node>

  <node pkg="viso2_slam" type="pose_optimizer" name="pose_optimizer" output="screen" >
    <param name="odom_frame_id"         value="$(arg world_frame_id)"/>
    <param name="base_link_frame_id"    value="$(arg base_link_frame_id)"/>
    <param name="gps_frame_id"          value="$(arg gps_frame_id)"/>
    <param name="pose_topic"            value="/stereo_slam/pose"/>
    <param name="info_topic"            value="/stereo_slam/info"/>
    <param name="publish_tf"            value="true"/>
    <param name="is_kitti"              value="$(arg is_kitti)"/>
    <param name="add_gps_noise"         value="$(arg add_gps_noise)"/>
  </node>

  <node pkg="viso2_slam" type="mapper" name="mapper" output="screen" >
    <param name="odom_frame_id"         value="$(arg world_frame_id)"/>
  </node>

  <!-- Trajectory server for gps, odometry, and kalman filtered gps -->
  <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="viso2_slam" output="screen" ns="gps" >
    <param name="/target_frame_name"    value="$(arg world_frame_id)" />
    <param name="/source_frame_name"    value="$(arg gps_frame_id)" />
    <param name="/trajectory_update_rate"   value="10.0" />
    <param name="/trajectory_publish_rate"  value="10.0" />
  </node>

  <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="kitti_ground_truth" output="screen" ns="odom" >
    <param name="/target_frame_name"    value="world" />
    <param name="/source_frame_name"    value="odom" />
    <param name="/trajectory_update_rate"   value="10.0" />
    <param name="/trajectory_publish_rate"  value="10.0" />
  </node>

  <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="gps_filtered" output="screen" ns="kalman" >
    <param name="/target_frame_name"    value="world" />
    <param name="/source_frame_name"    value="kalman" />
    <param name="/trajectory_update_rate"   value="10.0" />
    <param name="/trajectory_publish_rate"  value="10.0" />
  </node>

</launch>